src/alpaka/AlpakaCore/HistoContainer.h

```diff
@@ -0,0 +1,370 @@
+            assert((*off) > 0);
+            int32_t ih = off - offsets - 1;
+            assert(ih >= 0);
+            assert(ih < int(nh));
+            h->count(acc, v[i], ih);
+          }
+          endElementIdx += gridDimension;
+        }
+      }
+    };
+
+    struct fillFromVector {
+      template <typename T_Acc, typename Histo, typename T>
+      ALPAKA_FN_ACC void operator()(const T_Acc &acc,
+                                    Histo *__restrict__ h,
+                                    uint32_t nh,
+                                    T const *__restrict__ v,
+                                    uint32_t const *__restrict__ offsets) const {
+        const uint32_t nt = offsets[nh];
+        const uint32_t gridDimension(alpaka::workdiv::getWorkDiv<alpaka::Grid, alpaka::Elems>(acc)[0u]);
+        const auto &[firstElementIdxNoStride, endElementIdxNoStride] =
+            cms::alpakatools::element_global_index_range(acc, Vec1::all(nt));
+
+        uint32_t endElementIdx = endElementIdxNoStride[0u];
+        for (uint32_t threadIdx = firstElementIdxNoStride[0u]; threadIdx < nt; threadIdx += gridDimension) {
+          for (uint32_t i = threadIdx; i < std::min(endElementIdx, nt); ++i) {
+            auto off = alpaka_std::upper_bound(offsets, offsets + nh + 1, i);
+            assert((*off) > 0);
+            int32_t ih = off - offsets - 1;
+            assert(ih >= 0);
+            assert(ih < int(nh));
+            h->fill(acc, v[i], i, ih);
+          }
+          endElementIdx += gridDimension;
+        }
+      }
+    };
+
+    struct launchZero {
+      template <typename T_Acc, typename Histo>
+      ALPAKA_FN_ACC ALPAKA_FN_INLINE __attribute__((always_inline)) void operator()(const T_Acc &acc,
+                                                                                    Histo *__restrict__ h) const {
+        const auto &[firstElementIdxGlobal, endElementIdxGlobal] =
+            cms::alpakatools::element_global_index_range(acc, Vec1::all(Histo::totbins()));
+
+        for (uint32_t i = firstElementIdxGlobal[0u]; i < endElementIdxGlobal[0u]; ++i) {
+          h->off[i] = 0;
+        }
+      }
+    };
+
+    struct storePrefixScanWorkingSpace {
+      template <typename T_Acc, typename Histo>
+      ALPAKA_FN_ACC void operator()(const T_Acc &acc, Histo *__restrict__ h, const unsigned int nblocks) const {
+        h->psws = nblocks;
+      }
+    };
+
+    template <typename Histo>
+    ALPAKA_FN_HOST ALPAKA_FN_INLINE __attribute__((always_inline)) void launchFinalize(Histo *__restrict__ h,
+                                                                                       const DevAcc1 &device,
+                                                                                       Queue &queue) {
+      uint32_t *poff = (uint32_t *)((char *)(h) + offsetof(Histo, off));
+
+      const int num_items = Histo::totbins();
+
+      auto psum_dBuf = alpaka::mem::buf::alloc<uint32_t, Idx>(device, Vec1::all(num_items));
```

Comment:
Basically the prefixScan have slightly different interfaces in the CUDA and ALPAKA versions. In the Alpaka version, there is no `psws` argument.
I had looked at that specific point, and from what I remember, psws also exists in the Alpaka version, but as an internal variable (`ws `or sth).

The issue is that here we are actually interested in storing the `ws` value to Histo (even if I had looked up, and at least in the pixeltrack-standalone, `myHisto->psws` seems to never be used eventually).
To preserve the same Histo interface, I have, by safety, decided to set it later on (it actually very simply corresponds to the number of blocks). It is what `storePrefixScanWorkingSpace` is doing.
I have compared the Histo `off` and `psws` data mebers between CUDA and Alpaka versions and they are identical.

Reply:
I guess in ideal it would be better to have the exact same interface for a prefixScan with CUDA and Alpaka, but maybe there are reasons I miss.
In any case was planning to do a review of prefixScan for my understanding:
* There are uses of `alpaka::wait::wait(queue);` in between kernels, which I do not understand and maybe are not needed (only at the interface with CPU work should we need those).
* The size are made of the multiplication of the number of elements by sizeof(type), while with Alpaka only the number of elements should be specified (the multiplication with sizeof(type) is already done internally). Hence I guess the tests are done with 'too many' elements.
* I would like to check the CPU and TBB cases.

Reply:
> In the Alpaka version, there is no `psws` argument for doing a prefixScan.

Ok, so the situation is a bit more complicated.

The CUDA version takes `pc`, that is used for the "synchronization point" to know which of the threads belong to the "last block". Alpaka version does not need that because that synchronization is handled with splitting the code in two kernels.

The Alpaka version takes `psum` (well the `...FirstStep` doesn't seem to really need it), which in CUDA is handled as `extern __shared__ T psum[]` (does Alpaka support dynamic size array in shared memory?). 

In both cases the actual value of `pc`/`psum` is irrelevant outside of the kernels, so maybe we could re-purpose `HistoContainer::ppsws` for the `psum` here? I think we are not interested in the actual content of `psws`, it seems to be used only as temporary workspace (and e.g. in Kokkos version `HistoContainer` does not have `psws` member at all).

Reply:
> In any case was planning to do a review of prefixScan for my understanding:

Sounds like it could be a good idea.

> * There are calls to `alpaka::wait::wait(queue);` in between kernels, which I do not understand and maybe are not needed (we should only need those after a copy from device to host, when interested in what has been copied, or before a host function returns; well, whenever the host needs info from device).

I agree, we should not have `alpaka::wait::wait()` calls within these helper functions.

Reply:
>does Alpaka support dynamic size array in shared memory?

Yes.
- dynamic shared memory allocation with Alpaka:
`pBlockDynShared = alpaka::block::shared::dyn::getMem<TElem>(acc)`
- static shared memory allocation:
`pBlockStatShared = alpaka::block::shared::st::allocVar<TElem, __COUNTER__>(acc);`

Reply:
> I think we are not interested in the actual content of psws, it seems to be used only as temporary workspace

Yes I had only made a grep within the pixeltrack-standalone project, and there the Histogram data member called psws was never used.
Just doing a git grep in CMSSW, it appears it is never used there either.
I tend to think it is easier to just systematically keep the same interfaces, as it often eventually save issues dowsntream, but true that here, we can indeed also skip it, and simply just remove psws data member from Histo class.

Reply:
I'd still like to avoid this memory allocation (either by keeping `Histo::psws` or trying dynamic shared memory allocation in `multiBlockPrefixScanSecondStep`.

Reply:
>I'd still like to avoid this memory allocation (either by keeping Histo::psws or trying dynamic shared memory allocation in multiBlockPrefixScanSecondStep.

Yes doing that now.
Basically, since `pc` / `psum` are used internally only in `PrefixScan`, I do not really see why we should store `psum `inside Histo (even in place of `psws`). Additionally, `psws` was just a pointer to an `uint32_t`, whereas `psum `should be an array of `num_items` elements.

That's why I have just removed `psws` from Histo.
I think`psum` should be addressed internally in `PrefixScan', by indeed dynamic shared memory allocation, as for CUDA version. 
I was thinking doing a `prefixScan `related PR, also modifying the prefixScan call sites, but ok this point can also be addressed in this PR. Doing a commit ~now :)

Reply:
You're right, `psws` could not be used for `psum` because that is used as an array (whereas `pc` is used as an atomic counter).